quantities/concepts.hpp

```diff
@@ -14,36 +14,104 @@ namespace internal {
 using namespace base::_traits;
 using namespace quantities::_quantities;
 
+// TODO(egg): additive_group should subsume affine, but we use it there.
 template<typename G>
-concept additive_group = requires(G x, G y) {
+concept additive_group = requires(G x, G y, int n) {
   G{};
   { +x } -> std::same_as<G>;
   { -x } -> std::same_as<G>;
   { x + y } -> std::same_as<G>;
   { x - y } -> std::same_as<G>;
   { x += y } -> std::same_as<G&>;
   { x -= y } -> std::same_as<G&>;
+  // An abelian group is a ℤ-module; we require the corresponding operations.
+  { n * x } -> std::same_as<G>;
+  { x * n } -> std::same_as<G>;
+  { x *= n } -> std::same_as<G&>;
 };
 
+// A set acted upon simply transitively by an additive group.
+template<typename A>
+concept affine = requires(A x, A y) {
+  { x - y } -> additive_group;
+  { y + (x - y) } -> std::same_as<A>;
+  { y += (x - y) } -> std::same_as<A&>;
+  { y - (x - y) } -> std::same_as<A>;
+  { y -= (x - y) } -> std::same_as<A&>;
+};
+
+// A graded ring restricted to its homogeneous elements; multiplication can
+// alter the type, and addition is only defined between homogeneous elements.
+template<typename A>
+concept homogeneous_ring =
+    additive_group<A> && requires(A x, A y, A z) {
+      // Really, multiplication should return a homogeneous_ring.
+      { x * y } -> additive_group;
+      { (x * y) * z } -> additive_group;
+      { (x * y) * z } -> std::same_as<decltype(x * (y * z))>;
+};
+
+template<typename A>
+concept ring = homogeneous_ring<A> && requires(A x, A y) {
+  { x * y } -> std::same_as<A>;
+  { x *= y } -> std::same_as<A&>;
+};
+
+// TODO(egg): field should subsume homogeneous_field, but we use it in
+// homogeneous_field.
+
+template<typename K>
+concept field = ring<K> && !std::integral<K> && requires(K x, K y, K z) {
```

Comment:
It should apply to `std::complex<double>` too (but it doesn’t because that is missing multiplication by an integer).

It could apply to, _e.g._, the field of rational fractions over a field, if we decided to have those as well as polynomials.